### PR TITLE
[ISSUE 19] Pin and place important lists on top

### LINKED ISSUE #65: 19. As a user, I want to be able to mark a list as important and to have important lists on top
AC:
- There is a UX option available for me to pin/unpin important lists at the top
- The pinned lists are sorted in reverse add order.
- This is isolated to a per user bases.
- a11y checks have been done.
- Update the test to ensure coverage of this new functionality. 

## 🧗🏽 Challenge

- The audience for this change is users of the smart application list who frequently access specific lists and need a way to prioritize them.
- The user wants to be able to mark a list as important so that it is displayed at the top of their lists for easier access.
- This matters because it enhances user experience by allowing users to quickly navigate to the lists they use most often, making the application more user-centric.
- This change aligns with our broader product objective of improving usability and personalization in the application.
- We will accomplish this by adding a UX option for users to pin or unpin lists, sorting pinned lists in reverse order of their pinning date.

## ⚡️ Solution

- The solution involves implementing a feature that allows users to pin/unpin important lists at the top of their list view. This will include front-end adjustments for the UX.
**Possible Solution**: Each `SingleList` element will have an invisible pin icon and a state that tracks whether the list is being hovered over. Upon hovering, the pin icon becomes visible. The pin icon will have an `onClick` attribute that toggles the state, setting the list as pinned (`true` - marked as 'pinned') or unpinned (`false` - no longer marked as 'pinned'). When a list is pinned, the pin icon remains visible, and the list stays at the top of all lists until the pin icon is clicked again.

## 🎨 Designs

- Links to Figma files will be provided once the design is finalized. The design will highlight how pinned lists appear at the top and the option to pin/unpin them.

**Mockups**:

Hovered:
Image
Clicked:
Image
Pinned:
Image

## ✅ Acceptance Criteria

_Should describe the definition of done, not the technical how-to (technical details go in the Solution section)._

- [ ] The UX option to pin/unpin lists is available and functions correctly.
- [ ] Pinned lists are sorted in reverse order of their creation date.
- [ ] This functionality is isolated to each user's account.
- [ ] Accessibility (a11y) checks have been completed and validated.
- [ ] Tests are updated to ensure coverage of this new functionality.

## 🚫 Out of Scope

- This issue does not include changes to the overall list sorting mechanism beyond pinning functionality.
- Follow-on work may include enhancements such as notifications for pinned lists or additional sorting options. Related GitHub issues will be linked as they arise.

## 📝 Notes

- Additional context includes ensuring that the pinning feature is intuitive and easy to use. 
- Technical constraints include ensuring that the current data model can support the pinning functionality without significant refactoring.

### PR DESCRIPTION
**Pull request by kweeuhree**
## Description

This code proposes a custom hook that manages and sorts lists that were marked as important

- Created `useImportance` hook to manage a list of items and their importance status.
- Added functionality to toggle a list's importance and store the state in local storage.
     I am using localStorage to locally store references to important lists - like this we don't have to make any additional calls to firebase. 
- Implemented sorting logic to display important lists at the top based on the date they were pinned.
     I wanted to try and sort items the way that we originally wanted to implement urgency of items. I read a bit more about how sorting works in JavaScript and I think this time everything works correctly, and I am happy about this. 
- Added tooltips and aria-labels to the Pin icons.

**_Would love to change UI based on feedback_**

**Still working on tests**

## Related Issue

closes #65 

## Acceptance Criteria

- [x] There is a UX option available for me to pin/unpin important lists at the top
- [x] The pinned lists are sorted in reverse add order.
- [x] This is isolated to a per user bases.
- [x] a11y checks have been done.
- [x] Update the test to ensure coverage of this new functionality.

## Type of Changes

`enhancement`

### Before
Screenshot 2024-10-01 163249

### After
Screenshot 2024-10-14 143336

## Testing Steps / QA Criteria
pinlist-ezgif com-speed

1. ran npm start and signed in
2. clicked on lists to ensure they get added in the order they were pinned

### CODE REVIEW

File: src/components/SingleList.jsx [hunk 1]
```diff
@@ -1,13 +1,41 @@
...
+	const handleClick = () => {
 		setListPath(path);
-	}
+	};
+
+	const handleImportantList = () => {
+		setImportantList(path);
+	};
+
+	const pinStyle = isImportant ? 'visible' : !isHovered ? 'hidden' : 'visible';
```
Comment (kweeuhree) [inline src/components/SingleList.jsx:23]: pin is displayed based on whether it is: 
- important, 
- not hovered,
- hovered

File: src/components/SingleList.jsx [hunk 2]
```diff
@@ -1,13 +1,41 @@
...
+			<li
+				className={`SingleList ${isHovered && 'hovered'}`}
+				onMouseEnter={() => setIsHovered(true)}
+				onMouseLeave={() => setIsHovered(false)}
+			>
+				<PushPinIcon
+					fontSize="large"
+					className={pinStyle}
+					onClick={handleImportantList}
+				/>
```
Comment (kweeuhree) [inline src/components/SingleList.jsx:35]: the icon is handling setting selected list as important

File: src/utils/hooks.js [hunk 3]
```diff
@@ -19,3 +19,68 @@ export function useStateWithStorage(storageKey, initialValue) {
...
+ *  - {Array<Object>} sortedLists - An array of lists sorted by importance and date they were pinned.
+ *  - {Function} setImportantList - A function to toggle the importance of a list identified by its `path`.
+ *  - {Function} isListImportant - A function to check if a given list is marked as important.
+ */
+export function useImportance(lists) {
+	const [sortedLists, setSortedLists] = useState([]);
+	const [importantLists, setImportantLists] = useState(
+		() => JSON.parse(localStorage.getItem('importantLists')) || [],
+	);
+
```
Comment (kweeuhree) [inline src/utils/hooks.js:36]: use state to display real-time changes to local storage where we store an array of objects:
```
{ 
   path: string,
   datePinned: Date 
}
```
datePinned is needed to be able to display the lists in the reverse add order

File: src/utils/hooks.js [hunk 4]
```diff
@@ -19,3 +19,68 @@ export function useStateWithStorage(storageKey, initialValue) {
...
+			const bDate = getDatePinned(b.path);
+
+			return bDate - aDate;
+		} else if (aIsImportant) {
+			return -1;
+		} else if (bIsImportant) {
+			return 1;
+		}
+	};
+
```
Comment (kweeuhree) [inline src/utils/hooks.js:80]: I tried to use a ternary and it wasn't working at all!

**Comment by jendevelops** [comment]
I would consider the user experience here. Unless you have a perviously pinned list, or you hover over the lists a user would not know that you can pin a list.

One of the ways that we can visually indicate this ability is to have an "empty" or "outline only" version of the pin icon on every list item, and have the ones that are pinned filled in. The other option is to have the pin look more like a "button", and indicate pinned status through highlighting.


Also, it's important to remember that visual representation is not the only thing that's important for our app. For any icons, we should also make sure to try to add [alt tags and aria-labels](https://phuoc.ng/collection/this-vs-that/alt-vs-aria-label/) to make our visual indicators accessible. I linked a good article that highlights the differences between the 2 and how to use them together to make visual items on a page more accessible.

> 2024-10-0116-33-07-ezgif com-video-to-gif-converter

**Comment by kweeuhree** [comment]
> I would consider the user experience here. Unless you have a perviously pinned list, or you hover over the lists a user would not know that you can pin a list.
> 
> One of the ways that we can visually indicate this ability is to have an "empty" or "outline only" version of the pin icon on every list item, and have the ones that are pinned filled in. The other option is to have the pin look more like a "button", and indicate pinned status through highlighting.
> 
> Also, it's important to remember that visual representation is not the only thing that's important for our app. For any icons, we should also make sure to try to add [alt tags and aria-labels](https://phuoc.ng/collection/this-vs-that/alt-vs-aria-label/) to make our visual indicators accessible. I linked a good article that highlights the differences between the 2 and how to use them together to make visual items on a page more accessible.
> 
> > 2024-10-0116-33-07-ezgif com-video-to-gif-converter
> >       
> >         [
> >           
> >         
> >             2024-10-0116-33-07-ezgif com-video-to-gif-converter
> >           ](https://private-user-images.githubusercontent.com/160637676/372642760-ff2811d8-edb9-43c7-9750-38c5ce319daa.gif?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3Mjg4NTMwMTQsIm5iZiI6MTcyODg1MjcxNCwicGF0aCI6Ii8xNjA2Mzc2NzYvMzcyNjQyNzYwLWZmMjgxMWQ4LWVkYjktNDNjNy05NzUwLTM4YzVjZTMxOWRhYS5naWY_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQxMDEzJTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MTAxM1QyMDUxNTRaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT04YThmNDJlMDI4OGRhYTczOGM1N2Y3ZTAzNGZjMmM4NmI5NTQ2M2EyNzExMjkxYTZkNjRmOTk3Yzk0ZDFhZjk0JlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.CIvgL6-zwERSx6WTdpaxu_69zlK77SzurB-uvOM4nF8)
> >         
> >         
> >           
> >             
> >               
> >             
> >             
> >               
> >               
> >             
> >           
> >           [
> >             
> >               
> >             
> >           ](https://private-user-images.githubusercontent.com/160637676/372642760-ff2811d8-edb9-43c7-9750-38c5ce319daa.gif?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3Mjg4NTMwMTQsIm5iZiI6MTcyODg1MjcxNCwicGF0aCI6Ii8xNjA2Mzc2NzYvMzcyNjQyNzYwLWZmMjgxMWQ4LWVkYjktNDNjNy05NzUwLTM4YzVjZTMxOWRhYS5naWY_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQxMDEzJTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MTAxM1QyMDUxNTRaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT04YThmNDJlMDI4OGRhYTczOGM1N2Y3ZTAzNGZjMmM4NmI5NTQ2M2EyNzExMjkxYTZkNjRmOTk3Yzk0ZDFhZjk0JlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.CIvgL6-zwERSx6WTdpaxu_69zlK77SzurB-uvOM4nF8)

I agree, I will update this to be buttons for both delete and pin icons, and I will add aria-labels and alt tags! Thank you!

File: src/components/SingleList.jsx [hunk 5]
```diff
@@ -1,13 +1,63 @@
...
+		setImportantList(path);
+	};
+
+	const tooltipStyle = {
+		fontSize: '1.5rem',
+		marginBlockStart: '0',
+		marginBlockEnd: '0',
+	};
+
+	const tooltipTitle = isImportant ? 'Pinned' : 'Not pinned';
```
[the remark this answers is not in the document]
Comment (kweeuhree) [reply src/components/SingleList.jsx]: Combine tooltipTitle and iconButtonAreaLabel into importantStatusLabel!

`const importantStatusLabel = isImportant ? 'Unpin list' : 'Pin list';`

**Pull request merged**